### PR TITLE
16-pagination

### PR DESCRIPTION
**Pull request by steve9001**
https://trello.com/c/EYcshhzg/16-pagination-support-search-history

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -30,6 +30,33 @@ DEBUG environment variable, or use `node-fhir-client:*` to enable both.
...
+
+This approach is quicker and dirtier. Invoke the tests with
+
+```
+yarn test -- debug
+```
+
+Execution will be paused at the beginning, so at the `debug>` prompt enter `c` to continue execution. When a `debugger` statement is reached you will have the `debug>` prompt, but the execution context is not the context of the `debugger` statement.
+Therefore enter `repl` command to enter that context. When you're ready to proceed, use `Ctrl-C` to exit the debug repl, and you will be back at the `debug>` prompt, where you can enter `c` to continue.
+
```
Comment (foxtree) [inline README.md]: would it be better to recommend `.exit` to exit?

^ e.g.: it's the same way to exit a Node.js REPL.

Comment (foxtree) [reply README.md]: Recommending `.exit` also reinforces the other semantics and commands shared with Node.js:

image

Comment (foxtree) [reply README.md]: image

Comment (steve9001) [reply README.md]: I think I put `Ctrl-C` because it literally prints that instruction to exit. But I can update it

File: lib/client.js [hunk 2]
```diff
@@ -1,6 +1,7 @@
 const { URLSearchParams } = require('url');
 const { authFromCapability } = require('./smart');
 const HttpClient = require('./http-client');
+const { Pager } = require('./pager');
```
Comment (foxtree) [inline lib/client.js]: perhaps "Paginator" ?    (A "Pager" is a device popular in the 80's for calling back phone numbers sent to it)

File: lib/pager.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+   *
+   * @param {object} results a bundle result of a FHIR search
+   */
+  prevPage(results) {
+    const { url } = results.link.find(link => link.relation === 'previous');
+    return this.httpClient.get(url);
+  }
+}
+
+module.exports = { Pager };
```
Comment (foxtree) [inline lib/pager.js]: Given the name of the file, and the fact that this is the "main" / only thing this file does, perhaps:

`module.exports = Pager`

(analogous to the "default" export from standard ES2015+ ?)

Comment (bkaney) [reply lib/pager.js]: Can we have a discussion about this? I would like to know the best practices on `module.export`

File: lib/pager.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    return this.httpClient.get(url);
+  }
+
+  /**
+   * Return the previous page of results.
+   *
+   * @param {object} results a bundle result of a FHIR search
+   */
+  prevPage(results) {
+    const { url } = results.link.find(link => link.relation === 'previous');
```
Comment (Jammjammjamm) [inline lib/pager.js]: `prev` can also be used. https://www.iana.org/assignments/link-relations/link-relations.xhtml#link-relations-1

Comment (foxtree) [reply lib/pager.js]: perhaps:
```
  const { url } = results.link.find(link => link.relation.match(/^prev(ious)?$/));
```
...reusing pre-compiled regex might run more efficiently.

**Comment by bkaney** [comment]
This looks good. But one thing that occurred to me is tweaking the API so you could:

```js
client.nextPage(response);
```

instead of the less-Demeter-friendly:

```js
client.pager.nextPage(response);
```

File: test/client-test.js [hunk 5]
```diff
@@ -181,6 +190,45 @@ describe('Client', () => {
       });
     });
 
+    describe('#pagination', () => {
```
Comment (Jammjammjamm) [inline test/client-test.js]: Shouldn't include `#` here since `pagination` isn't an instance method.

**Pull request merged**